indico/queries/storage.py

```diff
@@ -133,3 +141,99 @@ def process_response(self, uploaded_files: List[dict]) -> List[str]:
 
 # Alias to ensure backwards compatibility
 UploadImages = CreateStorageURLs
+
+
+class GetDownloadUrl(GraphQLRequest):
+    """
+    Retrieve a signed url with a uri
+
+    Args:
+        uri (str): the storage uri prefixed by indico-file:///storage
+    Returns:
+        url (str): signed url of storage object
+    """
+
+    query = """
+        mutation($uri: String!) {
+            requestStorageDownloadUrl(uri: $uri) {
+                signedUrl
+            }
+        }
+    """
+
+    def __init__(self, *, uri: str):
+        uri = "indico-file:///storage" + uri
+        super().__init__(self.query, variables={"uri": uri})
+
+    def process_response(self, response: dict) -> str:
+        response = super().process_response(response)
+        return response["requestStorageDownloadUrl"]["signedUrl"]
+
+
+class GetUploadURL(GraphQLRequest):
+    """
+    Receive a signed url for the file being uploaded
+
+    """
+
+    query = """
+        mutation {
+            requestStorageUploadUrl {
+                signedUrl
+                relativePath
+            }
+        }
+    """
+
+    def __init__(self):
+        super().__init__(self.query)
+
+    def process_response(self, response):
+        response = super().process_response(response)
+
+        return {
+            "signed_url": response["requestStorageUploadUrl"]["signedUrl"],
+            "relative_path": response["requestStorageUploadUrl"]["relativePath"],
+        }
+
+
+class UploadSignedURL(HTTPRequest):
+    """
+    Upload files with a signed url
+
+    Args:
+        files (str): A list of local filepaths to upload_file.
+    """
+
+    def __init__(
+        self,
+        signed_url: str,
+        file: str,
+    ):
+        with open(file, "rb") as f:
+            super().__init__(
+                HTTPMethod.PUT, path=signed_url, data=f.read(), indico_url=False
+            )
```

Comment:
```suggestion
        super().__init__(
            HTTPMethod.PUT, path=signed_url, data=open(file, "rb"), indico_url=False
        )
```
so that it doesn't actually read the file into memory

Reply:
unfortunately it does not work. it doesn't accept a file handler - Jerry

Reply:
I was able to get it work with plain `requests` lib
```
>>> import requests
>>> url = "https://indico-data-blob-us-west-2.s3.amazonaws.com/blob/uploads/2/cbef3ad5-3a8f-4cc1-b15b-c786ffe86627?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAXWKQXD5FLCSIYW5P%2F20220930%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20220930T194322Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=***"
>>> requests.put(url, data=open("125.pdf", "rb"))
<Response [200]>
```
and I think the sdk uses `requests` behind the scene

Reply:
i receive this error when running it, probably how we make the request vs how requests does differs and throws this error?
`raceback (most recent call last):
  File "/home/ubuntu/indico/indico-client-python/test_blob.py", line 22, in <module>
    res = client.call(UploadSigned(jacob_file))
  File "/home/ubuntu/indico/indico-client-python/indico/client/client.py", line 77, in call
    return self._handle_request_chain(request)
  File "/home/ubuntu/indico/indico-client-python/indico/client/client.py", line 47, in _handle_request_chain
    response = self._http.execute_request(request)
  File "/home/ubuntu/indico/indico-client-python/indico/http/client.py", line 83, in execute_request
    self._make_request(
  File "/home/ubuntu/indico/indico-client-python/indico/http/client.py", line 149, in _make_request
    with self._handle_files(request_kwargs) as new_kwargs:
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/home/ubuntu/indico/indico-client-python/indico/http/client.py", line 99, in _handle_files
    new_kwargs = deepcopy(req_kwargs)
  File "/usr/lib/python3.10/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.10/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 161, in deepcopy
    rv = reductor(4)
TypeError: cannot pickle '_io.BufferedReader' object
`

Reply:
maybe try applying this patch in `indico/http/client.py` around line 97 and test again
```
+        # data could be a buffered reader for PUT
+        data = None
+        if isinstance(req_kwargs.get("data"), io.BufferedReader):
+            data = req_kwargs.pop("data")
+
         new_kwargs = deepcopy(req_kwargs)
 
+        if data is not None:
+            new_kwargs["data"] = data
```
you'll also need to add `import io`